[LV] Mask off possibly aliasing vector lanes

SamTebbs33 opened the pull request:
When vectorising a loop that uses loads and stores, those pointers with unknown bounds could overlap if their difference is less than the vector factor. For example, if address 20 is being stored to and address 23 is being loaded from, they overlap when the vector factor is 4 or higher. Currently LoopVectorize branches to a scalar loop in these cases with a runtime check. However if we construct a mask that disables the overlapping (aliasing) lanes then the vectorised loop can be safely entered, as long as the loads and stores are masked off.

This PR modifies the LoopVectorizer and VPlan to create such a mask and branch to the vector loop if the mask has active lanes.

huntergr-arm commented on clang/test/CodeGen/loop-alias-mask.c:
```diff
@@ -0,0 +1,404 @@
+// NOTE: Assertions have been autogenerated by utils/update_cc_test_checks.py UTC_ARGS: --version 4
```
We generally don't add source level tests for llvm optimizations; aiui clang's CodeGen test directory is to test lowering the AST to LLVM IR, not end-to-end.

We might want a different way of testing the feature from source that we could add to a buildbot.

huntergr-arm commented on llvm/include/llvm/Analysis/LoopAccessAnalysis.h:
```diff
@@ -448,6 +448,20 @@ struct PointerDiffInfo {
         NeedsFreeze(NeedsFreeze) {}
 };
 
+/// A pair of pointers that could overlap across a loop iteration.
```
If the analysis has been removed from LAA, then I suggest the struct can also be moved elsewhere. Maybe VPlan.h ?

SamTebbs33 replied:
That sounds good to me. I originally put it here so it was next to the related `PointerDiffInfo`

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -428,6 +431,85 @@ Value *VPInstruction::generatePerPart(VPTransformState &State, unsigned Part) {
+                          EltVT->getArrayNumElements();
+          else
+            ElementSize =
+                GEP->getSourceElementType()->getScalarSizeInBits() / 8;
```
Not sure if using the GEP source element type here is correct, should this use the accessed type from load/store which could be different?

SamTebbs33 replied:
I've managed to figure it out now and am using the `AccessSize` provided by `PointerDIffInfo`. Thanks for pointing out the flakiness here!

fhahn commented on llvm/test/Transforms/LoopVectorize/runtime-checks-difference.ll:
```diff
@@ -9,11 +9,11 @@ define void @same_step_and_size(ptr %a, ptr %b, i64 %n) {
 ; CHECK-NEXT:    [[B1:%.*]] = ptrtoint ptr [[B:%.*]] to i64
 ; CHECK-NEXT:    [[MIN_ITERS_CHECK:%.*]] = icmp ult i64 [[N:%.*]], 4
-; CHECK-NEXT:    br i1 [[MIN_ITERS_CHECK]], label %scalar.ph, label %vector.memcheck
+; CHECK-NEXT:    br i1 [[MIN_ITERS_CHECK]], label [[SCALAR_PH:%.*]], label [[VECTOR_MEMCHECK:%.*]]
```
Unrelated changes?

fhahn commented on llvm/test/Transforms/LoopVectorize/AArch64/whilewr-opt.ll:
```diff
@@ -0,0 +1,369 @@
+  %cmp11 = icmp sgt i32 %n, 0
+  br i1 %cmp11, label %for.body.preheader, label %for.cond.cleanup
+
+for.body.preheader:
```
I might be missing something but the input is already vectorized?

SamTebbs33 replied:
You're right, thanks for pointing that out. I've improved the test so that it shows the loop vectoriser's work.

SamTebbs33 commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9838,16 +9902,33 @@ bool LoopVectorizePass::processLoop(Loop *L) {
+  GeneratedRTChecks Checks(*PSE.getSE(), DT, LI, TTI, F->getDataLayout(),
+                           AddBranchWeights);
+
+  // VPlan needs the aliasing pointers as Values and not SCEVs, so expand them
```
Oh perfect, I didn't know that existed, thanks.

SamTebbs33 replied:
Done.

SamTebbs33 commented:
Ping.

huntergr-arm commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -2778,6 +2808,60 @@ void VPWidenPointerInductionRecipe::print(raw_ostream &O, const Twine &Indent,
+      }
+    }
+  }
+  assert(ElementSize > 0 && "Couldn't get element size from pointer");
```
This assert fires when compiling `MicroBenchmarks/ImageProcessing/Dilate/CMakeFiles/Dilate.dir/dilateKernel.c` from the test suite, so that needs fixing at least.

But I think we shouldn't be trying to figure out the element size at recipe execution time; it should be determined when trying to build the recipe and the value stored as part of the class. If we can't find the size then we stop trying to build the recipe and fall back to normal RT checks. Does that sound sensible to you?

SamTebbs33 replied:
Thanks Graham, I've had a deeper look and `PointerDiffInfo` provides an `AccessSize` which corresponds to what we want, so I'm using that now.

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -8520,7 +8561,29 @@ LoopVectorizationPlanner::tryToBuildVPlanWithVPRecipes(VFRange &Range) {
   bool HasNUW = Style == TailFoldingStyle::None;
-  addCanonicalIVRecipes(*Plan, Legal->getWidestInductionType(), HasNUW, DL);
+
+  VPValue *AliasMask = nullptr;
```
If possible to do so in a separate transform, that would preferable IMO as now the general VPlan construction has a bunch of added complexity for something that is quite specific.

A similar approach was taken for adding various options to control tail-folding (`addActiveLaneMask`, `tryAddExplicitVectorLength`). This also has the advantage of keeping the code responsible to generate the new construct at a single place, rather than spreading it across multiple places?

Having it as a separate transform also allows to generate VPlans with different strategies (e.g. with and without this new mask) and pick the best option based on cost.

SamTebbs33 commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2725,9 +2725,17 @@ class VPAliasLaneMaskRecipe : public VPSingleDefRecipe {
   /// Get the VPValue* for the pointer being stored to
   VPValue *getSinkValue() const { return getOperand(1); }
 
+private:
+  unsigned ElementSize;
+
```
This need to be moved above `public` or below `print`.

SamTebbs33 committed:
[LV] Mask off possibly aliasing vector lanes

    When vectorising a loop that uses loads and stores, those pointers could
    overlap if their difference is less than the vector factor. For example,
    if address 20 is being stored to and address 23 is being loaded from, they
    overlap when the vector factor is 4 or higher. Currently LoopVectorize
    branches to a scalar loop in these cases with a runtime check. Howver if
    we construct a mask that disables the overlapping (aliasing) lanes then
    the vectorised loop can be safely entered, as long as the loads and
    stores are masked off.

SamTebbs33 commented:
I've now fixed the comparison that's emitted since it should have been comparing to 0. See the Arm [WHILERW](https://developer.arm.com/documentation/ddi0602/2024-06/SVE-Instructions/WHILERW--While-free-of-read-after-write-conflicts-) instruction.

SamTebbs33 commented:
I've rebased this on top of my PR that adds an intrinsic since that's less fragile to match in the backend. So this should now be ready to have a look at.

MacDue commented on llvm/lib/Transforms/Vectorize/LoopVectorizationPlanner.h:
```diff
@@ -418,7 +418,13 @@ class LoopVectorizationPlanner {
   /// non-zero or all applicable candidate VFs otherwise. If vectorization and
   /// interleaving should be avoided up-front, no plans are generated.
-  void plan(ElementCount UserVF, unsigned UserIC);
+  /// RTChecks is a list of pointer pairs that should be checked for aliasing,
+  /// setting HasAliasMask to true in the case that an alias mask is generated
```
Outdated comment? Is this `DiffChecks` now?

MacDue commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -427,6 +428,29 @@ Value *VPInstruction::generate(VPTransformState &State) {
+  // Count the number of bits set in each lane and reduce the result to a scalar
+  case VPInstruction::PopCount: {
+    Value *Op = State.get(getOperand(0));
+    auto *VT = Op->getType();
```
nit: Spell out type if it's not present on the RHS.
```suggestion
    Type *VT = Op->getType();
```

SamTebbs33 replied:
Done.

MacDue commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -3235,6 +3263,36 @@ void VPWidenPointerInductionRecipe::print(raw_ostream &O, const Twine &Indent,
+  Value *SinkValue = State.get(getSinkValue(), true);
+  Value *SourceValue = State.get(getSourceValue(), true);
+
+  auto *Type = SinkValue->getType();
```
nit:
```suggestion
  Type *PtrType = SinkValue->getType();
```

SamTebbs33 replied:
Not needed thanks to rebase.

MacDue commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -3235,6 +3263,36 @@ void VPWidenPointerInductionRecipe::print(raw_ostream &O, const Twine &Indent,
+                                  VPSlotTracker &SlotTracker) const {
+  O << Indent << "EMIT ";
+  getVPSingleValue()->printAsOperand(O, SlotTracker);
+  O << " = alias lane mask ";
```
nit: These seem more commonly printed in all caps with hyphens. 
```suggestion
  O << " = ALIAS-LANE-MASK ";
```

SamTebbs33 replied:
Done.

MacDue commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -952,7 +952,6 @@ void VPlan::prepareToExecute(Value *TripCountV, Value *VectorTripCountV,
 
   IRBuilder<> Builder(State.CFG.PrevBB->getTerminator());
   // FIXME: Model VF * UF computation completely in VPlan.
-  assert(VFxUF.getNumUsers() && "VFxUF expected to always have users");
```
How does removing this assert relate to these changes?

MacDue commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1300,14 +1301,38 @@ static VPActiveLaneMaskPHIRecipe *addVPLaneMaskPhiAndUpdateExitBranch(
+  VPValue *AliasMask = nullptr;
+
+  for (auto C : RTChecks) {
+    // FIXME: How to pass this info back?
+    //    HasAliasMask = true;
```
This FIXME is a little unclear. Does it mean `HasAliasMask` should be set here but it's not?

SamTebbs33 replied:
The info is acutally being passed back so I can remove this FIXME. Done.

MacDue commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1416,14 +1466,14 @@ void VPlanTransforms::addActiveLaneMask(
       find_if(Plan.getCanonicalIV()->users(),
               [](VPUser *U) { return isa<VPWidenCanonicalIVRecipe>(U); });
-  assert(FoundWidenCanonicalIVUser &&
+  assert(FoundWidenCanonicalIVUser && *FoundWidenCanonicalIVUser &&
```
This looks a little odd. Doesn't `find_if` return an iterator? 
```suggestion
  auto IVUsers = Plan.getCanonicalIV()->users();
  /// ...
  assert(FoundWidenCanonicalIVUser != IVUsers.end() && "Must have widened canonical IV when tail folding!"); 
```

SamTebbs33 replied:
Done, thanks.

MacDue commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.h:
```diff
@@ -77,9 +77,13 @@ struct VPlanTransforms {
   /// creation) and instead it is handled using active-lane-mask. \p
   /// DataAndControlFlowWithoutRuntimeCheck implies \p
   /// UseActiveLaneMaskForControlFlow.
+  /// RTChecks refers to the pointer pairs that need aliasing elements to be
+  /// masked off each loop iteration.
```
No docs for PSE?

SamTebbs33 replied:
Added, let me know if anything about it should change.

MacDue commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -3073,6 +3075,56 @@ struct VPWidenStoreEVLRecipe final : public VPWidenMemoryRecipe {
+  bool WriteAfterRead;
+
+#if !defined(NDEBUG) || defined(LLVM_ENABLE_DUMP)
+  /// Print the recipe.
+  void print(raw_ostream &O, const Twine &Indent,
```
Would it be possible to add a test that prints the VPlan recipe with the alias mask?

SamTebbs33 replied:
Done.

MacDue commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1331,14 +1356,37 @@ static VPActiveLaneMaskPHIRecipe *addVPLaneMaskPhiAndUpdateExitBranch(
+    // And the alias mask so the iteration only processes non-aliasing lanes
+    Builder.setInsertPoint(CanonicalIVPHI->getParent(),
+                           CanonicalIVPHI->getParent()->getFirstNonPhi());
+    LaneMask = Builder.createNaryOp(Instruction::BinaryOps::And,
+                                    {LaneMaskPhi, AliasMask}, DL);
```
Do we know this AND won't be all-false?

SamTebbs33 replied:
We don't, and there's actually a case in the test suite that hangs because the mask is all-false. I'll start looking into a solution for that.

MacDue commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -3073,6 +3075,56 @@ struct VPWidenStoreEVLRecipe final : public VPWidenMemoryRecipe {
+// Given a pointer A that is being stored to, and pointer B that is being
+// read from, both with unknown lengths, create a mask that disables
+// elements which could overlap across a loop iteration. For example, if A
+// is X and B is X + 2 with VF being 4, only the final two elements of the
+// loaded vector can be stored since they don't overlap with the stored
+// vector. %b.vec = load %b ; = [s, t, u, v]
+// [...]
+// store %a, %b.vec ; only u and v can be stored as their addresses don't
+// overlap with %a + (VF - 1)
```
This is specifically RAW? Of something like:
```
store A[x]
load A[x + 2] 
```
Perhaps I'm muddled on what "final two elements" means, but isn't the first two elements store that is valid (so it won't overwrite the elements for the load)?

MacDue reviewed (commented):
A bunch of little comments (mostly just nitpicks from a pass over the PR) :slightly_smiling_face:

SamTebbs33 commented:
I've just pushed a new version that makes sure the popcount of the mask is greater than 0 before branching to the vector loop, otherwise it branches to the scalar loop as usual.

fhahn commented on llvm/include/llvm/Analysis/TargetTransformInfo.h:
```diff
@@ -1347,6 +1354,11 @@ class TargetTransformInfo {
+  /// \return true if a mask should be formed that disables lanes that could
+  /// alias between two pointers. The mask is created by the
+  /// loop_dependence_{war,raw}_mask intrinsics.
+  LLVM_ABI bool useSafeEltsMask(ElementCount VF) const;
```
Could we just check the cost of the new intrinsic using the existing hooks?

fhahn commented on llvm/lib/Transforms/Utils/LoopUtils.cpp:
```diff
@@ -2030,32 +2031,68 @@ Value *llvm::addDiffRuntimeChecks(
-      continue;
+    if (!VF.isScalar() && UseSafeEltsMask) {
+      Value *Sink = Expander.expandCodeFor(SinkStart, Ty, Loc);
+      Value *Src = Expander.expandCodeFor(SrcStart, Ty, Loc);
```
It doesn't look like there's much re-use going on here. Better to have a separate function than complicate the logic here?

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorizationPlanner.h:
```diff
@@ -479,7 +479,11 @@ class LoopVectorizationPlanner {
   /// non-zero or all applicable candidate VFs otherwise. If vectorization and
   /// interleaving should be avoided up-front, no plans are generated.
-  void plan(ElementCount UserVF, unsigned UserIC);
+  /// DiffChecks is a list of pointer pairs that should be checked for aliasing,
```
Better to add the new recipes at the same place where the runtime checks are attached to VPlan, rather than threading through the general planning stage: https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Vectorize/LoopVectorize.cpp#L9359C32-L9359C51

fhahn replied:
Yep that looks like a good first step. Would be good if updating the masks could also be done there (the PopCount introduction) and we could avoid making `addActiveLaneMask` more complex

SamTebbs33 commented:
Ping.

sdesmalen-arm commented on llvm/test/Transforms/LoopVectorize/AArch64/alias_mask.ll:
```diff
@@ -0,0 +1,350 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --filter-out-after "^middle.block:" --filter-out-after "^scalar.ph:" --version 4
```
This test currently fails.

sdesmalen-arm commented on llvm/lib/Transforms/Utils/LoopUtils.cpp:
```diff
@@ -2108,7 +2160,8 @@ Value *llvm::addDiffRuntimeChecks(
   DenseMap<std::pair<Value *, Value *>, Value *> SeenCompares;
-  for (const auto &[SrcStart, SinkStart, AccessSize, NeedsFreeze] : Checks) {
+  for (const auto &[SrcStart, SinkStart, AccessSize, NeedsFreeze,
+                    WriteAfterRead] : Checks) {
```
nit: `WriteAfterRead` is unused, so you can write `_` (because it's ignored).

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -4252,7 +4279,52 @@ void VPWidenPointerInductionRecipe::print(raw_ostream &O, const Twine &Indent,
+                                          Builder.getPtrTy());
+  Value *SinkAsPtr =
+      Builder.CreateCast(Instruction::IntToPtr, SinkValue, Builder.getPtrTy());
+  Value *AliasMask = Builder.CreateIntrinsic(
```
This basically lowers to a wide intrinsic, can this simply use VPWidenIntrinsicReicpe?

SamTebbs33 replied:
Figured it out now! I think I was getting the VPBuilder usage wrong so the recipes were being inserted in the wrong place.

fhahn commented on llvm/include/llvm/CodeGen/BasicTTIImpl.h:
```diff
@@ -2164,6 +2164,53 @@ class BasicTTIImplBase : public TargetTransformInfoImplCRTPBase<T> {
       // Otherwise, fallback to default scalarization cost.
       break;
     }
+    case Intrinsic::loop_dependence_raw_mask:
```
I think you should be able to move this to an independent PR, and also test this with dedicated cost-model tests.

SamTebbs33 committed:
Address review

SamTebbs33 committed:
[Analysis][AArch64] Add cost model for loop.dependence.{war/raw}.mask

This PR adds the cost model for the loop dependence mask intrinsics,
both for cases where they must be expanded and when they can be lowered
for AArch64.

SamTebbs33 commented on llvm/lib/Transforms/Utils/LoopUtils.cpp:
```diff
@@ -2094,6 +2094,58 @@ Value *llvm::addRuntimeChecks(
+
+  // Map to keep track of created compares, The key is the pair of operands for
+  // the compare, to allow detecting and re-using redundant compares.
+  DenseMap<std::pair<Value *, Value *>, Value *> SeenCompares;
```
No longer relevant now that this function has been removed, due to the new mask insertion code. Likewise for the other comments here.

SamTebbs33 commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -3394,6 +3396,59 @@ struct VPWidenStoreEVLRecipe final : public VPWidenMemoryRecipe {
+// [...]
+// store %b, %a.vec ; only s and t can be stored as their addresses don't
+// overlap with %a + (VF - 1)
+class VPAliasLaneMaskRecipe : public VPSingleDefRecipe {
```
I'm going to try to use `VPWidenIntrinsicRecipe` instead but have run into an issue.

SamTebbs33 committed:
Use VPWidenIntrinsicRecipe and clean up
